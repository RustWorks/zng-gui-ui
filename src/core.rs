pub mod animation;
pub mod app;
pub mod context;
pub mod event;
pub mod events;
pub mod focus;
pub mod font;
pub mod frame;
pub mod profiler;
pub mod types;
pub mod var;
pub mod window;

use crate::impl_ui_node;
use context::{LazyStateMap, WidgetContext};
use frame::FrameBuilder;
use types::{LayoutSize, WidgetId};

/// An Ui tree node.
pub trait UiNode: 'static {
    /// Called every time the node is plugged in an Ui tree.
    fn init(&mut self, ctx: &mut WidgetContext);

    /// Called every time the node is unplugged from an Ui tree.
    fn deinit(&mut self, ctx: &mut WidgetContext);

    /// Called every time a low pressure event update happens.
    ///
    /// # Event Pressure
    /// See [update_hp] for more information about event pressure rate.
    fn update(&mut self, ctx: &mut WidgetContext);

    /// Called every time a high pressure event update happens.
    ///
    /// # Event Pressure
    /// Some events occur alot more times then others, for performance reasons this
    /// event source may choose to be propagated in the this hight pressure lane.
    ///
    /// Event sources that are high pressure mention this in their documentation.
    fn update_hp(&mut self, ctx: &mut WidgetContext);

    /// Called every time a layout update is needed.
    ///
    /// # Arguments
    /// * `available_size`: The total available size for the node. Can contain positive infinity to
    /// indicate the parent will accommodate any size.
    ///
    /// # Return
    /// Must return the nodes desired size. Must not contain infinity or NaN.
    fn measure(&mut self, available_size: LayoutSize) -> LayoutSize;

    /// Called every time a layout update is needed, after [measure].
    ///
    /// # Arguments
    /// * `final_size`: The size the parent node reserved for the node. Must reposition its contents
    /// to fit this size. The value does not contain infinity or NaNs.
    fn arrange(&mut self, final_size: LayoutSize);

    /// Called every time a new frame must be rendered.
    ///
    /// # Arguments
    /// * `frame`: Contains the next frame draw instructions.
    fn render(&self, frame: &mut FrameBuilder);

    /// Box this component, unless it is already `Box<dyn UiNode>`.
    fn boxed(self) -> Box<dyn UiNode>
    where
        Self: Sized + 'static,
    {
        Box::new(self)
    }
}

#[impl_ui_node(delegate: self.as_ref(), delegate_mut: self.as_mut())]
impl UiNode for Box<dyn UiNode> {
    fn boxed(self) -> Box<dyn UiNode> {
        self
    }
}

struct Widget<T: UiNode> {
    id: WidgetId,
    state: LazyStateMap,
    child: T,
}

#[impl_ui_node(child)]
impl<T: UiNode> UiNode for Widget<T> {
    fn init(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.init(ctx));
    }

    fn deinit(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.deinit(ctx));
    }

    fn update(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update(ctx));
    }

    fn update_hp(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update_hp(ctx));
    }

    fn render(&self, frame: &mut FrameBuilder) {
        frame.push_widget(self.id, &self.child);
    }
}

/// Creates a widget bondary.
pub fn widget(id: WidgetId, child: impl UiNode) -> impl UiNode {
    Widget {
        id,
        state: LazyStateMap::default(),
        child,
    }
}
