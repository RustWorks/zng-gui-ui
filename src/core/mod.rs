#[macro_use]
mod macros;

mod focus;
mod font;
mod hittest;
mod keyboard;
mod mouse;
mod next_frame;
mod next_update;
mod ui;
mod ui_values;
mod window;

pub use focus::*;
pub use font::*;
pub use glutin::event::{ElementState, ModifiersState, MouseButton, ScanCode, VirtualKeyCode};
pub use glutin::window::CursorIcon;
pub use hittest::*;
pub use keyboard::*;
pub use mouse::*;
pub use next_frame::*;
pub use next_update::*;
pub use ui::*;
pub use ui_values::*;
pub use webrender::api::{
    ColorF, FontInstanceKey, FontKey, GradientStop, LayoutPoint, LayoutRect, LayoutSideOffsets, LayoutSize,
};
pub use webrender::euclid::{point2, size2};
pub use window::*;
