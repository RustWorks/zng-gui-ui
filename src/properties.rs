mod background;
mod border_impl;
mod context_var;
mod cursor_impl;
mod events;
mod focus;
mod layout;
mod ui_n;

pub use background::*;
pub use border_impl::*;
pub use context_var::*;
pub use cursor_impl::*;
pub use events::*;
pub use focus::*;
pub use layout::*;
pub use ui_n::*;
