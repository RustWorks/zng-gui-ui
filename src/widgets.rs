#[macro_use]
mod container_; // depends on container_
#[macro_use]
mod button_;

mod text;
mod view;

pub use button_::*;
pub use container_::*;
pub use text::*;
pub use view::*;
