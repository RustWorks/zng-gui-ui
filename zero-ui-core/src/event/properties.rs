use super::*;
use crate::{context::WidgetUpdates, handler::WidgetHandler, ui_node, widget_instance::UiNode};

#[doc(hidden)]
#[macro_export]
macro_rules! __event_property {
    (
        $(#[$on_event_attrs:meta])*
        $vis:vis fn $event:ident {
            event: $EVENT:path,
            args: $Args:path,
            filter: $filter:expr,
        }
    ) => { $crate::paste! {
        $(#[$on_event_attrs])*
        ///
        /// # Preview
        ///
        #[doc = "You can preview this event using [`on_pre_"$event "`](fn.on_pre_"$event ".html)."]
        /// Otherwise the handler is only called after the widget content has a chance of handling the event by stopping propagation.
        ///
        /// # Async
        ///
        /// You can use async event handlers with this property.
        #[$crate::property(EVENT, default( $crate::handler::hn!(|_, _|{}) ))]
        $vis fn [<on_ $event>](
            child: impl $crate::widget_instance::UiNode,
            handler: impl $crate::handler::WidgetHandler<$Args>,
        ) -> impl $crate::widget_instance::UiNode {
            $crate::event::on_event(child, $EVENT, $filter, handler)
        }

        #[doc = "Preview [`on_"$event "`](fn.on_"$event ".html) event."]
        ///
        /// # Preview
        ///
        /// Preview event properties call the handler before the main event property and before the widget content, if you stop
        /// the propagation of a preview event the main event handler is not called.
        ///
        /// # Async
        ///
        /// You can use async event handlers with this property, note that only the code before the fist `.await` is *preview*,
        /// subsequent code runs in widget updates.
        #[$crate::property(EVENT, default( $crate::handler::hn!(|_, _|{}) ))]
        $vis fn [<on_pre_ $event>](
            child: impl $crate::widget_instance::UiNode,
            handler: impl $crate::handler::WidgetHandler<$Args>,
        ) -> impl $crate::widget_instance::UiNode {
            $crate::event::on_pre_event(child, $EVENT, $filter, handler)
        }
    } };

    (
        $(#[$on_event_attrs:meta])*
        $vis:vis fn $event:ident {
            event: $EVENT:path,
            args: $Args:path,
        }
    ) => {
        $crate::__event_property! {
            $(#[$on_event_attrs])*
            $vis fn $event {
                event: $EVENT,
                args: $Args,
                filter: |ctx, args| true,
            }
        }
    };
}
///<span data-del-macro-root></span> Declare one or more event properties.
///
/// Each declaration expands to two properties `on_$event`, `on_pre_$event`.
/// The preview properties call [`on_pre_event`], the main event properties call [`on_event`].
///
/// # Examples
///
/// ```
/// # fn main() { }
/// # use zero_ui_core::event::{event_property, EventArgs};
/// # use zero_ui_core::keyboard::*;
/// event_property! {
///     /// on_key_input docs.
///     pub fn key_input {
///         event: KEY_INPUT_EVENT,
///         args: KeyInputArgs,
///         // default filter is |ctx, args| true,
///     }
///
///     pub(crate) fn key_down {
///         event: KEY_INPUT_EVENT,
///         args: KeyInputArgs,
///         // optional filter:
///         filter: |ctx, args| args.state == KeyState::Pressed,
///     }
/// }
/// ```
///
/// # Filter
///
/// App events are delivered to all `UiNode` inside all widgets in the [`UpdateDeliveryList`] and event subscribers list,
/// event properties can specialize further by defining a filter predicate.
///
/// The `filter` predicate is called if [`propagation`] is not stopped. It must return `true` if the event arguments
/// are relevant in the context of the widget and event property. If it returns `true` the `handler` closure is called.
/// See [`on_event`] and [`on_pre_event`] for more information.
///
/// If you don't provide a filter predicate the default always allows, so all app events targeting the widget and not already handled
/// are allowed by default.  Note that events that represent an *interaction* with the widget are send for both [`ENABLED`] and [`DISABLED`]
/// targets, event properties should probably distinguish if they fire on normal interactions vs on *disabled* interactions.
///
/// # Async
///
/// Async event handlers are supported by properties generated by this macro, but only the code before the first `.await` executes
/// in the event track, subsequent code runs in widget updates.
///
/// [`on_pre_event`]: crate::event::on_pre_event
/// [`on_event`]: crate::event::on_event
/// [`propagation`]: AnyEventArgs::propagation
/// [`ENABLED`]: crate::widget_info::Interactivity::ENABLED
/// [`DISABLED`]: crate::widget_info::Interactivity::DISABLED
#[macro_export]
macro_rules! event_property {
    ($(
        $(#[$on_event_attrs:meta])*
        $vis:vis fn $event:ident {
            event: $EVENT:path,
            args: $Args:path $(,
            filter: $filter:expr)? $(,)?
        }
    )+) => {$(
        $crate::__event_property! {
            $(#[$on_event_attrs])*
            $vis fn $event {
                event: $EVENT,
                args: $Args,
                $(filter: $filter,)?
            }
        }
    )+};
}
#[doc(inline)]
pub use crate::event_property;

/// Helper for declaring event properties.
///
/// This function is used by the [`event_property!`] macro.
///
/// # Filter
///
/// The `filter` predicate is called if [`propagation`] was not stopped. It must return `true` if the event arguments are
/// relevant in the context of the widget. If it returns `true` the `handler` closure is called. Note that events that represent
/// an *interaction* with the widget are send for both [`ENABLED`] and [`DISABLED`] targets, event properties should probably distinguish
/// if they fire on normal interactions vs on *disabled* interactions.
///
/// # Route
///
/// The event `handler` is called after the [`on_pre_event`] equivalent at the same context level. If the event
/// `filter` allows more then one widget and one widget contains the other, the `handler` is called on the inner widget first.
///
/// # Async
///
/// Async event handlers are called like normal, but code after the first `.await` only runs in subsequent updates. This means
/// that [`propagation`] must be stopped before the first `.await`, otherwise you are only signaling
/// other async tasks handling the same event, if they are monitoring the propagation handle.
///
/// [`propagation`]: AnyEventArgs::propagation
/// [`ENABLED`]: crate::widget_info::Interactivity::ENABLED
/// [`DISABLED`]: crate::widget_info::Interactivity::DISABLED
pub fn on_event<C, A, F, H>(child: C, event: Event<A>, filter: F, handler: H) -> impl UiNode
where
    C: UiNode,
    A: EventArgs,
    F: FnMut(&mut WidgetContext, &A) -> bool + 'static,
    H: WidgetHandler<A>,
{
    #[ui_node(struct OnEventNode<A: EventArgs> {
        child: impl UiNode,
        #[event] event: Event<A>,
        filter: impl FnMut(&mut WidgetContext, &A) -> bool + 'static,
        handler: impl WidgetHandler<A>,
    })]
    impl UiNode for OnEventNode {
        fn event(&mut self, ctx: &mut WidgetContext, update: &mut EventUpdate) {
            self.child.event(ctx, update);
            if let Some(args) = self.event.on(update) {
                if !args.propagation().is_stopped() && (self.filter)(ctx, args) {
                    self.handler.event(ctx, args);
                }
            }
        }

        fn update(&mut self, ctx: &mut WidgetContext, updates: &mut WidgetUpdates) {
            self.child.update(ctx, updates);
            self.handler.update(ctx);
        }
    }

    #[cfg(dyn_closure)]
    let filter: Box<dyn FnMut(&mut WidgetContext, &A) -> bool> = Box::new(filter);

    OnEventNode {
        child: child.cfg_boxed(),
        event,
        filter,
        handler: handler.cfg_boxed(),
    }
    .cfg_boxed()
}

/// Helper for declaring preview event properties.
///
/// This function is used by the [`event_property!`] macro.
///
/// # Filter
///
/// The `filter` predicate is called if [`propagation`] was not stopped. It must return `true` if the event arguments are
/// relevant in the context of the widget. If it returns `true` the `handler` closure is called. Note that events that represent
/// an *interaction* with the widget are send for both [`ENABLED`] and [`DISABLED`] targets, event properties should probably distinguish
/// if they fire on normal interactions vs on *disabled* interactions.
///
/// # Route
///
/// The event `handler` is called before the [`on_event`] equivalent at the same context level. If the event
/// `filter` allows more then one widget and one widget contains the other, the `handler` is called on the inner widget first.
///
/// # Async
///
/// Async event handlers are called like normal, but code after the first `.await` only runs in subsequent event updates. This means
/// that [`propagation`] must be stopped before the first `.await`, otherwise you are only signaling
/// other async tasks handling the same event, if they are monitoring the propagation handle.
///
/// [`propagation`]: AnyEventArgs::propagation
/// [`ENABLED`]: crate::widget_info::Interactivity::ENABLED
/// [`DISABLED`]: crate::widget_info::Interactivity::DISABLED
pub fn on_pre_event<C, A, F, H>(child: C, event: Event<A>, filter: F, handler: H) -> impl UiNode
where
    C: UiNode,
    A: EventArgs,
    F: FnMut(&mut WidgetContext, &A) -> bool + 'static,
    H: WidgetHandler<A>,
{
    #[ui_node(struct OnPreviewEventNode<A: EventArgs> {
        child: impl UiNode,
        #[event] event: Event<A>,
        filter: impl FnMut(&mut WidgetContext, &A) -> bool + 'static,
        handler: impl WidgetHandler<A>,
    })]
    impl UiNode for OnPreviewEventNode {
        fn event(&mut self, ctx: &mut WidgetContext, update: &mut EventUpdate) {
            if let Some(args) = self.event.on(update) {
                if !args.propagation().is_stopped() && (self.filter)(ctx, args) {
                    self.handler.event(ctx, args);
                }
            }
            self.child.event(ctx, update);
        }

        fn update(&mut self, ctx: &mut WidgetContext, updates: &mut WidgetUpdates) {
            self.handler.update(ctx);
            self.child.update(ctx, updates);
        }
    }

    #[cfg(dyn_closure)]
    let filter: Box<dyn FnMut(&mut WidgetContext, &A) -> bool> = Box::new(filter);

    OnPreviewEventNode {
        child: child.cfg_boxed(),
        event,
        filter,
        handler: handler.cfg_boxed(),
    }
    .cfg_boxed()
}
