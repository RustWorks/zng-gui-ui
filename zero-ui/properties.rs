//! Common widget properties.

mod util;
pub use util::*;

mod layout;
pub use layout::*;

mod visual;
pub use visual::*;

mod border_;
pub use border_::*;

mod capture_mouse_;
pub mod commands;
mod cursor_;
pub mod drag_move;
pub mod events;
pub mod filters;
pub mod focus;
pub mod states;
pub mod transform;

pub use capture_mouse_::*;
pub use cursor_::*;
