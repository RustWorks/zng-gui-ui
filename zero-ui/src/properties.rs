//! Common widget properties.

pub mod access;

pub mod inspector;

mod layout;
pub use layout::*;

mod visual;
pub use visual::*;

mod border_;
pub use border_::{border, border_align, corner_radius, corner_radius_fit, CornerRadiusFit};

pub mod mask;
pub use mask::{mask_align, mask_fit, mask_image, mask_mode};

pub mod commands;

pub mod data_context;
pub use data_context::{
    data, data_error, data_info, data_warn, get_data_error, get_data_error_txt, get_data_info, get_data_info_txt, get_data_warn,
    get_data_warn_txt, has_data_error, has_data_info, has_data_warn, DATA,
};

pub mod events;
pub mod filters;
pub mod focus;
pub mod states;
pub mod transform;

mod undo;
pub use undo::*;

mod capture;
pub use capture::*;

mod mouse;
pub use mouse::*;

mod lazy_;
pub use lazy_::{lazy, LazyMode};

pub use crate::widgets::{
    menu::context::{context_menu, context_menu_anchor, context_menu_fn, disabled_context_menu, disabled_context_menu_fn},
    tip::{
        disabled_tooltip, disabled_tooltip_fn, tooltip, tooltip_anchor, tooltip_context_capture, tooltip_delay, tooltip_duration,
        tooltip_fn, tooltip_interval,
    },
};

pub use crate::core::widget_base::{enabled, hit_test_mode, interactive};
