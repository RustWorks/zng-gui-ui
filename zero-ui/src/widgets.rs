//! Common widgets.

pub mod layouts;

pub mod focusable_mix;
pub mod undo_mix;

pub mod ansi_text;
pub use ansi_text::AnsiText;

pub mod button;
pub use button::Button;

pub mod checkerboard;
pub use checkerboard::Checkerboard;

mod container;
pub use container::Container;

mod flood;
pub use flood::flood;

pub mod gradient;
pub use gradient::{conic_gradient, gradient, linear_gradient, radial_gradient};

pub mod image;
pub use image::Image;

pub mod icon;
pub use icon::Icon;

pub mod markdown;
pub use markdown::Markdown;

pub mod rule_line;
pub use rule_line::{hr::Hr, vr::Vr, RuleLine};

pub mod scroll;
pub use scroll::Scroll;

pub mod switch;
pub use switch::Switch;

pub mod text;
pub use text::{Em, Strong, Text};

pub mod text_input;
pub use text_input::TextInput;

pub mod tip;
pub use tip::Tip;

pub mod toggle;
pub use toggle::Toggle;

pub mod style;
pub use style::Style;

mod view;
pub use view::*;

pub mod window;
pub use window::Window;

/// Minimal widget.
///
/// You can use this to create a quick new custom widget that is only used in one code place and can be created entirely
/// by properties and `when` conditions.
#[crate::core::widget($crate::widgets::Wgt)]
pub struct Wgt(crate::core::widget_base::WidgetBase);
