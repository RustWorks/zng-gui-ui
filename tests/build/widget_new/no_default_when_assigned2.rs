use zero_ui::properties::margin;
use zero_ui::widgets::blank;
use zero_ui::core::{property, UiNode, var::IntoVar};

#[property(context)]
pub fn my_property(child: impl UiNode, a: impl IntoVar<u32>) -> impl UiNode {
    let _ = a;
    child
}

fn main() {
    let _ = blank! {
        margin = 10;
        when self.my_property == 1 {
            margin = 20;
        }
    };
}
