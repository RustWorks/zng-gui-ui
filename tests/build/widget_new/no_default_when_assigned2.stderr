error: property `my_property` is not assigned and has no default value
  --> $DIR/no_default_when_assigned2.rs:14:19
   |
14 |         when self.my_property == 1 {
   |                   ^^^^^^^^^^^
